### ISSUE TITLE
Support ctc + n-gram LM decoding

### ISSUE THREAD

**Issue by zhaoyukoon** [wenet-e2e/wenet#775]
Currently, Wenet support fst-based LM decoding, which is a copy of kaldi and relies on external open-fst package.

I try to implement ctc with n-gram decoding in wenet, which contains following steps:

- [x]  1. Convert arpa-based language model file to binary trie format, which accelerates the LM model loading.
- [x]  2. C​ompute LM probability given a sentence, which is identical with SRILM impl.
- [x]  3. ​Implement CTC decoding with language model with [cube-pruning algorithm](https://aclanthology.org/P07-1019.pdf) in `ctc_ngram_beam_search.[h/cc]`.

In total, around 600 lines of code is added with compatibility remained.  On aishell1 test set, decoding with big LM(wenet speech text + aishell train) can improve WER from 4.70 to 4.10, which is close to FST-based result. Besides, LM integration brings 20% extra decoding time compared to prefix decoding.

However, two issues remained:

1. Only character-based LM is supported. I will try to implement word-based LM in future. When LM is trained on aishell train, no gain can be achieved on char-based LM (4.70 -> 4.66), which is bad than FST result.
2. The relation between attention+rescoring with LM decoding is unclear, `ctc_ngram_beam_search` returns nbest-result list from ctc+lm decoding directly.

In conclusion, I try to add ngram module to wenet, so as to remove dependency of kaldi and open-fst. Do you have any opinion or suggestion on this work?  Thanks!

**Comment by robin1001** [comment]
1. We have tried before, character-based LM always shows poor performance than word-based LM.
2. In my opinion, it's clear, no matter how we get the n-best candidates.

**Comment by robin1001** [comment]
Here is the design ideas of WeNet LM, which shows why we use n-gram + WFST based LM.
https://wenet-e2e.github.io/wenet/lm.html

**Comment by zhaoyukoon** [comment]
The motivation of this issue is to remove the dependency on kaldi and openFST. There should be no difference between WFST and trie approaches  under WER or latency.  [DeepSpeech](https://github.com/PaddlePaddle/DeepSpeech/blob/master/decoders/swig/ctc_beam_search_decoder.cpp) has its own decoding implement.

I used to work on statistical machine translation(SMT), where n-gram LM is used in a trie way rather than WFST. One typical SMT open-source system is [moses](https://github.com/moses-smt/mosesdecoder).

I am wondering whether it is neccesary to add trie approach to wenet's main repo.

**Comment by HalFTeen** [comment]
WFST is more popular in decoder included LM，because rich theory and mature tool(openfst). Trie LM is more difficult for many people because the way to search LM score need implement by yourself.

**Comment by zhaoyukoon** [comment]
> the way to search LM score need implement by yourself

Actually only 100 lines of code is need to implement lm score calculation.

I agree that WFST is more popular in speech field.  However, trie-based LM is much simpler and consistent with neural LM, e.g. [feedforward](https://aclanthology.org/D13-1140.pdf) or [recurrent](https://aclanthology.org/D13-1106.pdf).

**Comment by misaka23** [comment]
@zhaoyukoon Hi,this code base on beam search or prefix beam search？

**Comment by misaka23** [comment]
@zhaoyukoon Hi,this code based on beam search or prefix beam search？

**Comment by zhaoyukoon** [comment]
> @zhaoyukoon Hi,this code based on beam search or prefix beam search？

beam search, not prefix beam search.
Actually, it is same as to ctc wfst beam search, with except of usage of ngram LM.

**Comment by hellomotor** [comment]
I've tried to intergrate n-gram LM decoding,  looks promissing.

**Comment by Yymax-max** [comment]
your job seem nice ，look foward to

**Comment by xingchensong** [comment]
Will hold this PR for anyone else who has encountered and interested in the same question.
